Addition of the GPU Activity Agent to steer GPU frequency based upon GPU Compute Activity

lhlawson opened the pull request:
- Fixes #1906  
- Fixes #1908

- Relies upon changes made in PR #1914 #1917 #1945 and should not be integrated prior to review those PRs
- Agent operates best when using Fefficient information obtained from GPU frequency sweeps.  Experiments providing GPU frequency sweeps through the integration infrastructure are introduced in PR #2024, #2031.  These may rely upon PR #2010 

This PR introduces an agent that modulates the GPU frequency within a set of user specified bounds based upon the GPU_COMPUTE_ACTIVITY signal.  For NVIDIA systems this is the SM_ACTIVE signal.  The user may also provide an optional input 'phi' which can bias the agents frequency selection towards energy efficiency (phi of 1.0) or performance (phi of 0.0).  The default behavior if phi = 0.5

Primary changes:
- addition of GEOPM_CXX_MAN_GPUActivityAgent.3.rst
- addition of geopm_agent_gpu_activity.7.rst
- src/GPUActivityAgent.cpp
- src/GPUActivityAgent.hpp
- test/GPUActivityAgentTest.cpp
- integration/experiment/energy_efficiency/Makefile.mk
- integration/experiment/energy_efficiency/gpu_activity.py
- integration/experiment/energy_efficiency/run_gpu_activity_parres_dgemm.py
- integration/experiment/energy_efficiency/run_gpu_activity_parres_nstream.py
- integration/frequency_sweep/gen_gpu_activity_policy_recommendation.py

Additional tasks to consider:
- [x] Integration tests based upon the run_gpu_activity_parress_dgemm/stream.py files
- [x] Experiment scripts to find the Fe (freq_efficient) policy input based upon the GPU frequency sweep of parres dgemm
- [x] Experiment script updates (energy_efficiency/gpu_activity.py) to automatically provide the 'optimal' config instead of a simple phi sweep with default HPE sx40 values.

lhlawson committed:
Added GPU Activity Agent

Signed-off-by: lowren.h.lawson@intel.com <lowren.h.lawson@intel.com>

lhlawson committed:
First round of GPU Activity Agent Cleanup

Signed-off-by: Lowren Lawson <lowren.h.lawson@intel.com>

lhlawson committed:
Added initial GPU Activity Agent Documentation

Signed-off-by: Lowren Lawson <lowren.h.lawson@intel.com>

lhlawson committed:
GPU Activity Agent Unit Tests

Signed-off-by: Lowren Lawson <lowren.h.lawson@intel.com>

lhlawson committed:
Updated Agent to use 100ms DCGM rate, Utilization to offset short phase behavior, and added ROI tracking

Signed-off-by: lowren.h.lawson@intel.com <lowren.h.lawson@intel.com>

lhlawson committed:
Fixed DCGM signal/control query mixup and control value scale

Signed-off-by: lowren.h.lawson@intel.com <lowren.h.lawson@intel.com>

lhlawson committed:
Updated GPU Activity Agent documentation

Signed-off-by: lowren.h.lawson@intel.com <lowren.h.lawson@intel.com>

lhlawson commented on integration/experiment/energy_efficiency/gpu_activity.py:
```diff
@@ -0,0 +1,142 @@
+    sys_max = mach.frequency_max()
+    sys_sticker = mach.frequency_sticker()
+    config_list = [
+                     {"ACCELERATOR_FREQ_MAX": 1530000000,
```
Update util.py to query frequency min, max, etc for GPUs

lhlawson committed:
Added gpu_activity to TestAgent list

Signed-off-by: lowren.h.lawson@intel.com <lowren.h.lawson@intel.com>

lhlawson committed:
removed unnecessary makefile include of MockNVMLDevicePool

Signed-off-by: lowren.h.lawson@intel.com <lowren.h.lawson@intel.com>

lhlawson committed:
added gpu activity parres nstream files

Signed-off-by: lowren.h.lawson@intel.com <lowren.h.lawson@intel.com>

dannosliwcd commented on ronn/geopm_agent_gpu_activity.7.ronn:
```diff
@@ -0,0 +1,161 @@
+of 0.0) run at the Fe frequency, high activity regions (compute activity of 1.0)
+run at the Fmax frequency, and regions in between the extremes run at a frequency
+selected via
+F = = Fe + (Fmax - Fe) * ACCELERATOR_COMPUTE_ACTIVITY/UTILIZATION_ACCELERATOR.
```
```suggestion
F = Fe + (Fmax - Fe) * ACCELERATOR_COMPUTE_ACTIVITY/UTILIZATION_ACCELERATOR.
```

dannosliwcd commented on ronn/geopm_agent_gpu_activity.7.ronn:
```diff
@@ -0,0 +1,161 @@
+selected via
+F = = Fe + (Fmax - Fe) * ACCELERATOR_COMPUTE_ACTIVITY/UTILIZATION_ACCELERATOR.
+UTILIZATION_ACCELERATOR is used to scale the ACCELERATOR_COMPUTE_ACTIVITY in order
+to scale frequency selection with the percentage of time a kernel is runnning on
```
```suggestion
to scale frequency selection with the percentage of time a kernel is running on
```

dannosliwcd commented on ronn/geopm_agent_gpu_activity.7.ronn:
```diff
@@ -0,0 +1,161 @@
+* `Agent Name`:
+  Set the `--geopm-agent` launch option or `GEOPM_AGENT` environment
+  variable to "energy_efficient" and the Controller will select the
+  EnergyEfficientAgent for its control handler.  See
```
Can probably emit the --geopm-agent stuff, or make it specific to this agent

dannosliwcd commented on ronn/geopm_agent_gpu_activity.7.ronn:
```diff
@@ -0,0 +1,161 @@
+                     0.0 and 1.0. If NAN is passed, it will use 0.5 by default.
+
+* `Agent Sample Definitions`:
+  N/A
```
Looks like the agent adds all of its traced signals as samples, outputting differenced values for energy and raw values for the rest.

lhlawson committed:
Addressed Review Feedback on GPU CA

Signed-off-by: lowren.h.lawson@intel.com <lowren.h.lawson@intel.com>

lhlawson committed:
update GPUActivity documentation from ronn to sphinx format

Signed-off-by: lowren.h.lawson@intel.com <lowren.h.lawson@intel.com>

lhlawson committed:
Moved agent from general format to specific approach

Signed-off-by: lowren.h.lawson@intel.com <lowren.h.lawson@intel.com>

lhlawson committed:
removed GPUActivityAgent ronn files

Signed-off-by: lowren.h.lawson@intel.com <lowren.h.lawson@intel.com>

lhlawson committed:
Header exemptions, documentation updates for GPU-CA

Signed-off-by: lowren.h.lawson@intel.com <lowren.h.lawson@intel.com>

lhlawson committed:
Updates to policy validation and documentation

Signed-off-by: lowren.h.lawson@intel.com <lowren.h.lawson@intel.com>

bgeltz commented on service/docs/source/GEOPM_CXX_MAN_GPUActivityAgent.3.rst:
```diff
@@ -0,0 +1,98 @@
+
+#include `<geopm/GPUActivityAgent.hpp> <https://github.com/geopm/geopm/blob/dev/src/GPUActivityAgent.hpp>`_\
+
+Link with ``-lgeopm`` **(MPI)** or ``-lgeopmpolicy`` **(non-MPI)**
```
Please add that this requires either `--enable-nvml` or `--enable-levelzero` at configure time.

lhlawson replied:
Added that it requires --enable-nvml and --enable-dcgm on NVIDIA systems.  
Agent and/or IO Group updates are likely required for LevelZero systems, I'll file an issue/feature request for that update (including documentation)

bgeltz commented on service/docs/source/reference.rst:
```diff
@@ -79,6 +79,7 @@ GEOPM Manual Pages
    GEOPM_CXX_MAN_EnergyEfficientAgent.3
    GEOPM_CXX_MAN_EnergyEfficientRegion.3
    GEOPM_CXX_MAN_Exception.3
+   GEOPM_CXX_MAN_GPUActivityAgent.3
```
Perhaps this should also be added to the `GEOPM HPC Runtime Agents` section below.

bgeltz commented on service/docs/source/reference.rst:
```diff
@@ -79,6 +79,7 @@ GEOPM Manual Pages
    GEOPM_CXX_MAN_EnergyEfficientAgent.3
    GEOPM_CXX_MAN_EnergyEfficientRegion.3
    GEOPM_CXX_MAN_Exception.3
```
I'm also just noticing that the FrequencyMapAgent is not listed here.

bgeltz commented on service/docs/source/GEOPM_CXX_MAN_GPUActivityAgent.3.rst:
```diff
@@ -0,0 +1,98 @@
+Link with ``-lgeopm`` **(MPI)** or ``-lgeopmpolicy`` **(non-MPI)**
+
+
+.. code-block:: c++
```
I went looking at the other docu to see if we listed out all the agent methods, and I don't see that elseware.  I question the value of this block.

To me it's just confusing.  I can always go read the header file if I'm curious.  My vote would be to remove this code block section.

bgeltz replied:
I see it's in there for the EE agent now.  I think it should be removed from there as well.

lhlawson replied:
Agreed.

bgeltz commented on service/docs/source/GEOPM_CXX_MAN_GPUActivityAgent.3.rst:
```diff
@@ -0,0 +1,98 @@
+-------------
+
+**TODO**
+
```
```suggestion
```

bgeltz commented on service/docs/source/geopm_agent_gpu_activity.7.rst:
```diff
@@ -0,0 +1,158 @@
+
+The goal of this Agent is to save GPU energy by scaling GPU frequency based upon
+the compute activity of each GPU as provided by the GPU_COMPUTE_ACTIVITY
+signal and modified by the GPU_UTILIZATIOR signal.
```
```suggestion
signal and modified by the GPU_UTILIZATION signal.
```

bgeltz commented on service/docs/source/geopm_agent_gpu_activity.7.rst:
```diff
@@ -0,0 +1,158 @@
+is provided via the policy as ``GPU_FREQ_MAX`` and Fe is provided via
+the policy as ``GPU_FREQ_EFFICIENT``.  Low activity regions (compute activity
+of 0.0) run at the Fe frequency, high activity regions (compute activity of 1.0)
+run at the Fmax frequency, and regions in between the extremes run at a frequency
```
```suggestion
run at the Fmax frequency, and regions in between the extremes run at a frequency (F)
```

bgeltz commented on service/docs/source/geopm_agent_gpu_activity.7.rst:
```diff
@@ -0,0 +1,158 @@
+the policy as ``GPU_FREQ_EFFICIENT``.  Low activity regions (compute activity
+of 0.0) run at the Fe frequency, high activity regions (compute activity of 1.0)
+run at the Fmax frequency, and regions in between the extremes run at a frequency
+selected using the equation
```
```suggestion
selected using the equation:
```

bgeltz commented on service/docs/source/geopm_agent_gpu_activity.7.rst:
```diff
@@ -0,0 +1,158 @@
+the compute activity of each GPU as provided by the GPU_COMPUTE_ACTIVITY
+signal and modified by the GPU_UTILIZATIOR signal.
+
+The **GPUActivityAgent** scales frequency in the range of Fe to Fmax, where Fmax
```
```suggestion
The **GPUActivityAgent** scales frequency in the range of ``Fe`` to ``Fmax``, where ``Fmax``
```

I think it makes it more readable when the variables are offset from the rest of the text.  I'm in favor of doing this everywhere in this doc.

bgeltz commented on service/docs/source/geopm_agent_gpu_activity.7.rst:
```diff
@@ -0,0 +1,158 @@
+by the policy.  In the extreme case (phi of 1.0) Fmax will be lowered to Fe.
+
+For NVIDIA based systems the GPUActivityAgent attempts to set the
+DCGM::FIELD_UPDATE_RATE to 100ms, DCGM::MAX_STORAGE_TIME to 1s, and DCGM::MAX_SAMPLES
```
```suggestion
``DCGM::FIELD_UPDATE_RATE`` to 100ms, ``DCGM::MAX_STORAGE_TIME`` to 1s, and ``DCGM::MAX_SAMPLES``
```

bgeltz commented on service/docs/source/geopm_agent_gpu_activity.7.rst:
```diff
@@ -0,0 +1,158 @@
+
+For NVIDIA based systems the GPUActivityAgent attempts to set the
+DCGM::FIELD_UPDATE_RATE to 100ms, DCGM::MAX_STORAGE_TIME to 1s, and DCGM::MAX_SAMPLES
+to 100.  While the DCGM documentation indicates that users should 'generally' query
```
```suggestion
to 100.  While the DCGM documentation indicates that users should generally query
```

bgeltz commented on service/docs/source/geopm_agent_gpu_activity.7.rst:
```diff
@@ -0,0 +1,158 @@
+to 100.  While the DCGM documentation indicates that users should 'generally' query
+no faster than 100ms, the interface allows for setting the polling rate in the
+microsecond range. If the agent is intended to be used with workloads that exhibit
+extremely short phase behavioor a 1ms polling rate can be used.
```
```suggestion
extremely short phase behavior a 1ms polling rate can be used.
```

bgeltz commented on service/docs/source/geopm_agent_gpu_activity.7.rst:
```diff
@@ -0,0 +1,158 @@
+     of Interest (ROI) where ROI is determined as the
+     first sample of GPU activity to the last sample of GPU
+     activity.
+  ``GPU # Active Region Start Time``\ :
```
```suggestion
  ``GPU # Active Region Stop Time``\ :
```

bgeltz commented on service/docs/source/conf.py:
```diff
@@ -140,6 +140,7 @@ def get_description(rst_file_name):
     "GEOPM_CXX_MAN_EnergyEfficientAgent.3",
     "GEOPM_CXX_MAN_EnergyEfficientRegion.3",
     "GEOPM_CXX_MAN_Exception.3",
```
Surprised Fmap is missing.  Not this PR's problem though.

bgeltz commented on service/docs/source/geopm.7.rst:
```diff
@@ -268,87 +270,87 @@ GEOPM ENVIRONMENT VARIABLES
 
 
-* 
+*
```
Looks like this is all just dangling whitespace.

Was this here for a reason @cmcantalupo?

bgeltz commented on src/GPUActivityAgent.hpp:
```diff
@@ -0,0 +1,134 @@
+/*
+ * Copyright (c) 2015, 2016, 2017, 2018, 2019, Intel Corporation
```
```suggestion
 * Copyright (c) 2015 - 2022, Intel Corporation
```

Please make this change everywhere that it's needed in this PR.

bgeltz commented on src/GPUActivityAgent.hpp:
```diff
@@ -0,0 +1,134 @@
+
+#include "Agent.hpp"
+#include "geopm_time.h"
+#include "config.h"
```
Please move `config.h` to the cpp.  Make it the first include in the list.

In general take a look at:
https://github.com/geopm/geopm/blob/dev/service/src/PlatformIO.cpp#L33

I'm trying to do this in all the files I edit when I see it.  I'm trying to follow the ordering pattern in deea1df1025c56390cfaaf16a138f4897f4b0bbd:

    config.h, PIOImp.hpp,
    C includes, STL C++ includes,
    GEOPM public includes,
    GEOPM private C includes, GEOPM private C++ includes.

bgeltz commented on src/GPUActivityAgent.hpp:
```diff
@@ -0,0 +1,134 @@
+#ifndef GPUACTIVITYAGENT_HPP_INCLUDE
+#define GPUACTIVITYAGENT_HPP_INCLUDE
+
+#include <vector>
```
```suggestion
#include <vector>
#include <functional>
```

Needed for `std::function`

bgeltz commented on src/GPUActivityAgent.cpp:
```diff
@@ -0,0 +1,460 @@
+                                         domain_idx), NAN});
+        }
+
+        auto all_names = m_platform_io.control_names();
+        if (all_names.find("DCGM::FIELD_UPDATE_RATE") != all_names.end()) {
+            // DCGM documentation indicates that users should query no faster than 100ms
+            // even though the interface allows for setting the polling rate in the us range.
+            // In practice reducing below the 100ms value has proven functional, but should only
+            // be attempted if there is a proven need to catch short phase behavior that cannot
+            // be accomplished with the default settings.
+            m_platform_io.write_control("DCGM::FIELD_UPDATE_RATE", GEOPM_DOMAIN_BOARD, 0, 0.1); //100ms
+            m_platform_io.write_control("DCGM::MAX_STORAGE_TIME", GEOPM_DOMAIN_BOARD, 0, 1);
+            m_platform_io.write_control("DCGM::MAX_SAMPLES", GEOPM_DOMAIN_BOARD, 0, 100);
+        }
```
Thinking about this code block when we're using LevelZero and not NVML...

I'm wondering if this code that effectively initializes the specific IOGroup (DCGM in this case) actually belongs inside the IOGroup itself.  It struck me as odd that the agent here needs to understand the specifics of the IOGroup that is currently compiled in.

lhlawson replied:
For this agent the current default values used by the DCGMDevicePool in dev should suffice, but the user may want to change them if their workload has shorter phases.

We could remove this code if we both assume that a user will do a `geompwrite DCGM::FIELD_UPDATE_RATE board 0 <better value>` prior to their if there's a known better value _and the DCGM setting is persistent for the duration of their time on the system_.  Whenever the dcgmWatchFields function is called in the DCGMDevicePool it needs to provide the three values being written, and prior to the GEOPM service I don't believe there a way to modify those outside the agent that would persist until the agent ran.

bgeltz commented on src/GPUActivityAgent.cpp:
```diff
@@ -0,0 +1,460 @@
+            //Perf Biased.  Scale F_efficient up to F_max based upon phi value
+            //Active region phi usage
+            f_efficient = std::min(f_max, f_efficient + f_range * (0.5-phi) / 0.5);
+        }
```
What about when `phi = 0.5`?

lhlawson replied:
When phi=0.5 we allow the agent to steer frequency in the range of Fe to Fmax without modifying either Fe or Fmax.

bgeltz commented on src/GPUActivityAgent.cpp:
```diff
@@ -0,0 +1,460 @@
+
+    // Distribute incoming policy to children
+    void GPUActivityAgent::split_policy(const std::vector<double>& in_policy,
+                                    std::vector<std::vector<double> >& out_policy)
```
```suggestion
                                        std::vector<std::vector<double> >& out_policy)
```

bgeltz commented on src/GPUActivityAgent.cpp:
```diff
@@ -0,0 +1,460 @@
+
+            if (!std::isnan(gpu_compute_activity)) {
+                // Boundary Checking
+                gpu_compute_activity = std::min(gpu_compute_activity, 1.0);
```
Should this be an error condition?  When `gpu_compute_activity` is > 1.0?

lhlawson replied:
gpu_compute_activity is provided by DCGM through the nv-hostengine process.  Depending on how nv-hostengine is configured and how noisy the data hw & nv-hostengine return there's the possibility the values could exceed 1.0.  When seen in testing this behavior was transient so the approach of simply clamping to the maximum expected value of 1.0 was used vs GEOPM throwing an error due to bad data inputs.

The end result of gpu_compute_activity > 1.0 is the algorithm selecting the maximum frequency (being performance biased) provided by the user/admin.

bgeltz commented on src/GPUActivityAgent.cpp:
```diff
@@ -0,0 +1,460 @@
+                //
+                // This approach assumes the efficient frequency is suitable as both a
+                // baseline for active regions and and inactive regions. This is generally
+                // true of the efficient frequency is low power enough at idle due to clock
```
```suggestion
                // true when the efficient frequency consumes low power at idle due to clock
```

bgeltz commented on src/GPUActivityAgent.cpp:
```diff
@@ -0,0 +1,460 @@
+            }
+
+            // Frequency clamping
+            f_request = std::min(f_request, m_f_max);
```
Might want to add a `ifdef GEOPM_DBUG` with some prints that your frequency was clamped one way or the other unless it's easy to determine from the reports and traces.

lhlawson replied:
Added a "GPU Clipped Frequency Request" value to the report when GEOPM_DEBUG is defined.  I also moved the per GPU Region of interest start & stop time to the GEOPM_DEBUG case, as per GPU region time is enough for most non-debug cases.

lhlawson committed:
Addressed Review Feedback

Signed-off-by: lowren.h.lawson@intel.com <lowren.h.lawson@intel.com>

lhlawson committed:
Test include update after rebase on dev

Signed-off-by: lowren.h.lawson@intel.com <lowren.h.lawson@intel.com>

lhlawson committed:
documentation fixups

Signed-off-by: lowren.h.lawson@intel.com <lowren.h.lawson@intel.com>

lhlawson committed:
Updated agent to take wait time/sample period as a policy input

Signed-off-by: lowren.h.lawson@intel.com <lowren.h.lawson@intel.com>

lhlawson committed:
Added clipped frequency reporting to debug case.  Moved region start/stop time to debug reporting case

Signed-off-by: lowren.h.lawson@intel.com <lowren.h.lawson@intel.com>

lhlawson commented:
Latest updates include a SAMPLE_PERIOD policy value that allows the end user to specify sample period/wait time.  This should help address any control loop timing issues that stem from vastly different GPU counts or GPU signal & control access time varying from system to system.

lhlawson committed:
Updated header of gpu_activity python scripts

Signed-off-by: lowren.h.lawson@intel.com <lowren.h.lawson@intel.com>

lhlawson committed:
removed geopm_internal.h header

Signed-off-by: lowren.h.lawson@intel.com <lowren.h.lawson@intel.com>

lhlawson committed:
License Header update

Signed-off-by: lowren.h.lawson@intel.com <lowren.h.lawson@intel.com>

lhlawson committed:
Updated GPUActivityAgent to use GEOPM_DOMAIN_GPU

Signed-off-by: lowren.h.lawson@intel.com <lowren.h.lawson@intel.com>

lhlawson committed:
Added policy recommendation script for GPU-CA

Signed-off-by: lowren.h.lawson@intel.com <lowren.h.lawson@intel.com>

lhlawson committed:
Fixed missing file in frequency_sweep extra_dist

Signed-off-by: lowren.h.lawson@intel.com <lowren.h.lawson@intel.com>

bgeltz commented on src/GPUActivityAgent.cpp:
```diff
@@ -0,0 +1,464 @@
+        // value provided by the policy may not be valid.  In this case approximating
+        // f_efficient as midway between F_min and F_max is reasonable.
+        if (std::isnan(in_policy[M_POLICY_GPU_FREQ_EFFICIENT])) {
+            in_policy[M_POLICY_GPU_FREQ_EFFICIENT] = (in_policy[M_POLICY_GPU_FREQ_MAX]
+                                                             + gpu_min_freq) / 2;
```
```suggestion
            in_policy[M_POLICY_GPU_FREQ_EFFICIENT] = (in_policy[M_POLICY_GPU_FREQ_MAX] + gpu_min_freq) / 2;
```

or align it one space in from the open paren on the line above:
```suggestion
            in_policy[M_POLICY_GPU_FREQ_EFFICIENT] = (in_policy[M_POLICY_GPU_FREQ_MAX]
                                                      + gpu_min_freq) / 2;
```

bgeltz commented on src/GPUActivityAgent.cpp:
```diff
@@ -0,0 +1,464 @@
+        std::vector<std::pair<std::string, std::string> > result;
+
+        result.push_back({"GPU Frequency Requests", std::to_string(m_gpu_frequency_requests)});
+#ifdef GEOPM_DEBUG
```
Using a debug build is going to slow down the overall runtime.  I'm not sure how big the impact would be on this Agent without doing some testing. It seems that if you're trying to study something around clipping, you ought to use a release build to ensure things run at full speed.

Why not just include this all the time?

lhlawson replied:
As long as it doesn't add too much to the report including it all the time seems reasonable.

bgeltz commented on src/GPUActivityAgent.cpp:
```diff
@@ -0,0 +1,464 @@
+                              " Active Region Energy", std::to_string(energy_stop - energy_start)});
+            result.push_back({"GPU " + std::to_string(domain_idx) +
+                              " Active Region Time", std::to_string(region_stop - region_start)});
+#ifdef GEOPM_DEBUG
```
Again, these region times might look pretty different between a debug and release build.  If that matters, I'd just do this all the time.

lhlawson replied:
These I'd omit, as ACTIVE REGION TIME provides most of the relevant information.

lhlawson committed:
Addresssed review feedback on wsp and debug report values

Signed-off-by: lowren.h.lawson@intel.com <lowren.h.lawson@intel.com>

lhlawson commented on service/docs/source/geopm_agent_gpu_activity.7.rst:
```diff
@@ -0,0 +1,158 @@
+*
+  **Agent Policy Definitions**:
+
+      The ``Fe``, ``Fmax``, and  and maximum frequency are policy values.
```
```suggestion
      ``Fe``, ``Fmax``, and ``phi`` are policy values.
```

lhlawson commented on service/docs/source/geopm_agent_gpu_activity.7.rst:
```diff
@@ -0,0 +1,158 @@
+  ``GPU_PHI``\ :
+      The performance bias knob.  The value must be between
+      0.0 and 1.0. If NAN is passed, it will use 0.5 by default.
+
```
Add Sample Rate

lhlawson commented on integration/experiment/energy_efficiency/gpu_activity.py:
```diff
@@ -0,0 +1,128 @@
+    sys_max = mach.frequency_max()
+    sys_sticker = mach.frequency_sticker()
+    config_list = [
+                     {"GPU_FREQ_MAX": float("NAN"),
```
Add command line argument to provide GPU_FREQ_MAX and GPU_FREQ_EFFICIENT.   Generate this list from phi 0-1 in 0.1 increments instead of hard-coding

lhlawson committed:
updated GPU Activity experiment script to generate policy via list comprehension

Signed-off-by: lowren.h.lawson@intel.com <lowren.h.lawson@intel.com>

lhlawson committed:
Rename to GEOPM 2.0 signal conventions

Signed-off-by: lowren.h.lawson@intel.com <lowren.h.lawson@intel.com>

lhlawson commented on service/docs/source/geopm_agent_gpu_activity.7.rst:
```diff
@@ -0,0 +1,146 @@
+SEE ALSO
+--------
+
+:doc:`geopm(7) <geopm.7.html>`\ ,
+:doc:`geopm_agent_monitor(7) <geopm_agent_monitor.7.html>`\ ,
+:doc:`geopm_agent_energy_efficient(7) <geopm_agent_energy_efficient.7.html>`\ ,
+:doc:`geopm::Agent(3) <GEOPM_CXX_MAN_Agent.3.html>`\ ,
+:doc:`geopm_agent_c(3) <geopm_agent_c.3.html>`\ ,
+:doc:`geopm_prof_c(3) <geopm_prof_c.3.html>`\ ,
+:doc:`geopmagent(1) <geopmagent.1.html>`\ ,
+:doc:`geopmlaunch(1) <geopmlaunch.1.html>`
```
```suggestion
:doc:`geopm(7) <geopm.7>`\ ,
:doc:`geopm_agent_monitor(7) <geopm_agent_monitor.7>`\ ,
:doc:`geopm_agent_energy_efficient(7) <geopm_agent_energy_efficient.7>`\ ,
:doc:`geopm::Agent(3) <GEOPM_CXX_MAN_Agent.3>`\ ,
:doc:`geopm_agent_c(3) <geopm_agent_c.3>`\ ,
:doc:`geopm_prof_c(3) <geopm_prof_c.3>`\ ,
:doc:`geopmagent(1) <geopmagent.1>`\ ,
:doc:`geopmlaunch(1) <geopmlaunch.1>`
```

lhlawson commented on src/GPUActivityAgent.cpp:
```diff
@@ -0,0 +1,453 @@
+    // Describes expected policies to be provided by the resource manager or user
+    std::vector<std::string> GPUActivityAgent::policy_names(void)
+    {
+        return {"GPU_FREQ_MAX", "GPU_FREQ_EFFICIENT", "GPU_PHI", "SAMPLE_PERIOD"};
```
```suggestion
        return {"GPU_FREQ_MAX", "GPU_FREQ_EFFICIENT", "GPU_PHI"};
```
Post 2.0 Agents are expected to use a more general geopmlaunch option to set sample period/control rate.  SAMPLE_PERIOD should be removed from the agent and supporting collateral

lhlawson committed:
Addressed review feedback: removed DCGM setting & sample period, fixed docs

Signed-off-by: Lowren Lawson <lowren.h.lawson@intel.com>

lhlawson committed:
Updated GPU Activity Agent to use GPU min/max control alias

Signed-off-by: lowren.h.lawson@intel.com <lowren.h.lawson@intel.com>

lhlawson committed:
Updated unit tests for min/max controls

Signed-off-by: lowren.h.lawson@intel.com <lowren.h.lawson@intel.com>

lhlawson committed:
Added gpu_activity_agent inference test

Signed-off-by: lowren.h.lawson@intel.com <lowren.h.lawson@intel.com>

lhlawson committed:
Moved gpu activity agent policy recommendation script to upcoming gpu_frequency_sweep folder

Signed-off-by: lowren.h.lawson@intel.com <lowren.h.lawson@intel.com>

lhlawson committed:
Updated integration infrastructure with skip util for workload not present

Signed-off-by: lowren.h.lawson@intel.com <lowren.h.lawson@intel.com>

lhlawson committed:
Updated gpu activity agent and util.py based on offline review

Signed-off-by: lowren.h.lawson@intel.com <lowren.h.lawson@intel.com>

lhlawson committed:
Re-added handling for finding energy efficient frequency of multiple trials

Signed-off-by: lowren.h.lawson@intel.com <lowren.h.lawson@intel.com>

lhlawson committed:
added phi selection to gpu_activity.py to shorten integration test time if required

Signed-off-by: lowren.h.lawson@intel.com <lowren.h.lawson@intel.com>

lhlawson commented on integration/test/test_gpu_activity_agent.py:
```diff
@@ -0,0 +1,165 @@
+
+import geopmpy.agent
+import geopmpy.io
+from yaml import load
+try:
+    from yaml import CSafeLoader as SafeLoader
+except ImportError:
+    from yaml import SafeLoader
+from collections import defaultdict
```
```suggestion
```

lhlawson commented on integration/test/test_gpu_activity_agent.py:
```diff
@@ -0,0 +1,165 @@
+            parres_init_setup=None,
+            parres_exp_setup=None,
+            parres_teardown=None,
+            parres_args=None,
```
It's beneficial to increase the parres nstream input size due to the relatively short runtime with the default settings, but not necessarily required.

lhlawson committed:
Removed unneeded imports from integration test

Signed-off-by: lowren.h.lawson@intel.com <lowren.h.lawson@intel.com>

avilcheslopez commented on src/GPUActivityAgent.cpp:
```diff
@@ -0,0 +1,436 @@
+    // Validate incoming policy and configure default policy requests.
+    void GPUActivityAgent::validate_policy(std::vector<double> &in_policy) const
+    {
+        assert(in_policy.size() == M_NUM_POLICY);
```
GEOPM_DEBUG_ASSERT

avilcheslopez commented on src/GPUActivityAgent.cpp:
```diff
@@ -0,0 +1,436 @@
+    void GPUActivityAgent::split_policy(const std::vector<double>& in_policy,
+                                        std::vector<std::vector<double> >& out_policy)
+    {
+        assert(in_policy.size() == M_NUM_POLICY);
```
GEOPM_DEBUG_ASSERT

avilcheslopez commented on src/GPUActivityAgent.cpp:
```diff
@@ -0,0 +1,436 @@
+
+    void GPUActivityAgent::adjust_platform(const std::vector<double>& in_policy)
+    {
+        assert(in_policy.size() == M_NUM_POLICY);
```
GEOPM_DEBUG_ASSERT

avilcheslopez commented on src/GPUActivityAgent.cpp:
```diff
@@ -0,0 +1,436 @@
+                // being done (such as SM_ACTIVE for NVIDIA GPUs).
+                //
+                // The compute activity is scaled by the GPU Utilization, to help
+                // address the issues that come from workloads have short phases that are
```
...workloads ^that^ have...

avilcheslopez commented on src/GPUActivityAgent.cpp:
```diff
@@ -0,0 +1,436 @@
+                // frequency from system characterization.
+                //
+                // This approach assumes the efficient frequency is suitable as both a
+                // baseline for active regions and and inactive regions. This is generally
```
and and

avilcheslopez commented on src/GPUActivityAgent.cpp:
```diff
@@ -0,0 +1,436 @@
+            gpu_freq_request.push_back(f_request);
+        }
+
+        if (!gpu_freq_request.empty()) {
```
Why is this check necessary? Is it possible M_NUM_GPU could be 0? Will this agent load if there are no GPUs available?

lhlawson replied:
It's no longer necessary.  In previous revisions it was possible to hit a condition where the gpu_freq_request had not been initialized.  It may now be removed.

avilcheslopez commented on src/GPUActivityAgent.cpp:
```diff
@@ -0,0 +1,436 @@
+    // Read signals from the platform and calculate samples to be sent up
+    void GPUActivityAgent::sample_platform(std::vector<double> &out_sample)
+    {
+        assert(out_sample.size() == M_NUM_SAMPLE);
```
GEOPM_DEBUG_ASSERT

avilcheslopez reviewed (commented):
Just leaving some minor comments...

avilcheslopez committed:
Use GEOPM_DEBUG_ASSERT, fix minor typos in code doc

Signed-off-by: Alejandro Vilches <alejandro.vilches@intel.com>

avilcheslopez commented on src/GPUActivityAgent.cpp:
```diff
@@ -0,0 +1,440 @@
+        geopm_time(&m_last_wait);
+    }
+
+    // Push signals and controls for future batch read/write
```
This comment seems to apply to the init_platform_io() method.

avilcheslopez commented on src/GPUActivityAgent.cpp:
```diff
@@ -0,0 +1,440 @@
+    // Indicate whether to send the policy down to children
+    bool GPUActivityAgent::do_send_policy(void) const
+    {
+        return true;
```
Should this be false? Policy needs to be sent once in the beginning.

avilcheslopez commented on src/GPUActivityAgent.cpp:
```diff
@@ -0,0 +1,440 @@
+            m_gpu_active_energy_stop.push_back(0.0);
+        }
+
+        if (level == 0) {
+            init_platform_io();
+        }
```
Is level == 0 the root? Does this agent need to act differently for children and parent?

lhlawson committed:
Addressed review feedback: comment fixup, removed unused checks, addressed agent level, fixed integration test options

Signed-off-by: lowren.h.lawson@intel.com <lowren.h.lawson@intel.com>

lhlawson committed:
Integration test refactor

Signed-off-by: lowren.h.lawson@intel.com <lowren.h.lawson@intel.com>

avilcheslopez commented on src/GPUActivityAgent.cpp:
```diff
@@ -0,0 +1,439 @@
+
+        m_time = {m_platform_io.push_signal("TIME", GEOPM_DOMAIN_BOARD, 0), NAN};
+
+        for (int domain_idx = 0; domain_idx < M_NUM_GPU; ++domain_idx) {
```
Since this "for" loop iterates over the same values as the one above, it seems like they could be merged.

lhlawson commented:
- [x] Add info on system characterization to docs (what workload, how, etc)
- [x] Add info on generating policy
- [x] JSON policy requires double quotes for gen_pol*

lhlawson committed:
Review feedback: root/leaf fix, json policy fix, documentation

Signed-off-by: lowren.h.lawson@intel.com <lowren.h.lawson@intel.com>

avilcheslopez commented on service/docs/source/geopm_agent_gpu_activity.7.rst:
```diff
@@ -0,0 +1,161 @@
+-----------
+
+The goal of this Agent is to save GPU energy by scaling GPU frequency based upon
+the compute activity of each GPU as provided by the GPU_COMPUTE_ACTIVITY
```
GPU_CORE_ACTIVITY

avilcheslopez commented on service/docs/source/geopm_agent_gpu_activity.7.rst:
```diff
@@ -0,0 +1,161 @@
+run at the ``Fmax`` frequency, and regions in between the extremes run at a frequency (F)
+selected using the equation:
+
+``F = Fe + (Fmax - Fe) * GPU_COMPUTE_ACTIVITY/GPU_UTILIZATION``
```
GPU_CORE_ACTIVITY

avilcheslopez commented on service/docs/source/geopm_agent_gpu_activity.7.rst:
```diff
@@ -0,0 +1,161 @@
+
+``F = Fe + (Fmax - Fe) * GPU_COMPUTE_ACTIVITY/GPU_UTILIZATION``
+
+``GPU_UTILIZATION`` is used to scale the ``GPU_COMPUTE_ACTIVITY`` in order
```
GPU_CORE_ACTIVITY

avilcheslopez commented on integration/experiment/energy_efficiency/gpu_activity.py:
```diff
@@ -0,0 +1,89 @@
+                                enable_traces=args.enable_traces,
+                                enable_profile_traces=args.enable_profile_traces)
+
+def main(app_conf, **defaults):
+    parser = argparse.ArgumentParser()
+    setup_run_args(parser)
+    parser.set_defaults(**defaults)
+    args, extra_args = parser.parse_known_args()
+    launch(app_conf=app_conf, args=args,
+           experiment_cli_args=extra_args)
```
Note: minor opportunity to refactor, as all experiment dir main modules have the exact same ``main()``.

avilcheslopez commented on test/GPUActivityAgentTest.cpp:
```diff
@@ -0,0 +1,434 @@
+        std::unique_ptr<GPUActivityAgent> m_agent;
+        std::vector<double> m_default_policy;
+        size_t m_num_policy;
+        double m_freq_min;
+        double m_freq_max;
```
Could & should be const

avilcheslopez commented on test/GPUActivityAgentTest.cpp:
```diff
@@ -0,0 +1,434 @@
+    policy[FREQ_MAX] = m_freq_max;
+    policy[FREQ_EFFICIENT] = m_freq_max / 2;
+    policy[PHI] = 0.1;
+    m_agent->validate_policy(policy);
```
Expect that it doesn't throw. Same applies to all calls of validate_policy(), really.

avilcheslopez commented on test/GPUActivityAgentTest.cpp:
```diff
@@ -0,0 +1,434 @@
+    EXPECT_CALL(*m_platform_io, read_signal("GPU_CORE_FREQUENCY_MAX_AVAIL", _, _)).WillRepeatedly(
+                Return(m_freq_max));
+    EXPECT_CALL(*m_platform_topo, num_domain(GEOPM_DOMAIN_BOARD)).WillRepeatedly(Return(M_NUM_BOARD));
+    std::vector<double> policy;
+    policy = m_default_policy;
```
Merge these two lines: std::vector<double> policy = m_default_policy

avilcheslopez commented on test/GPUActivityAgentTest.cpp:
```diff
@@ -0,0 +1,434 @@
+    EXPECT_CALL(*m_platform_topo, num_domain(GEOPM_DOMAIN_BOARD)).WillRepeatedly(Return(M_NUM_BOARD));
+
+    const std::vector<double> empty(m_num_policy, NAN);
+    std::vector<double> policy;
+
+    // default policy is accepted
+    // load default policy
+    policy = m_default_policy;
```
Merge these two lines: std::vector policy = m_default_policy

avilcheslopez commented on test/GPUActivityAgentTest.cpp:
```diff
@@ -0,0 +1,434 @@
+    EXPECT_CALL(*m_platform_io, read_signal("GPU_CORE_FREQUENCY_MAX_AVAIL", _, _)).WillRepeatedly(
+                Return(m_freq_max));
+    EXPECT_CALL(*m_platform_topo, num_domain(GEOPM_DOMAIN_BOARD)).WillRepeatedly(Return(M_NUM_BOARD));
+    std::vector<double> policy;
+    policy = m_default_policy;
```
Merge these two lines: std::vector policy = m_default_policy

avilcheslopez commented on test/GPUActivityAgentTest.cpp:
```diff
@@ -0,0 +1,434 @@
+    EXPECT_CALL(*m_platform_io, read_signal("GPU_CORE_FREQUENCY_MAX_AVAIL", _, _)).WillRepeatedly(
+                Return(m_freq_max));
+    EXPECT_CALL(*m_platform_topo, num_domain(GEOPM_DOMAIN_BOARD)).WillRepeatedly(Return(M_NUM_BOARD));
+    std::vector<double> policy;
+    policy = m_default_policy;
```
Merge these two lines: std::vector policy = m_default_policy

avilcheslopez commented on test/GPUActivityAgentTest.cpp:
```diff
@@ -0,0 +1,434 @@
+    EXPECT_CALL(*m_platform_io, read_signal("GPU_CORE_FREQUENCY_MAX_AVAIL", _, _)).WillRepeatedly(
+                Return(m_freq_max));
+    EXPECT_CALL(*m_platform_topo, num_domain(GEOPM_DOMAIN_BOARD)).WillRepeatedly(Return(M_NUM_BOARD));
+    std::vector<double> policy;
+    policy = m_default_policy;
```
Merge these two lines: std::vector policy = m_default_policy

avilcheslopez commented on test/GPUActivityAgentTest.cpp:
```diff
@@ -0,0 +1,434 @@
+    EXPECT_CALL(*m_platform_io, read_signal("GPU_CORE_FREQUENCY_MAX_AVAIL", _, _)).WillRepeatedly(
+                Return(m_freq_max));
+    EXPECT_CALL(*m_platform_topo, num_domain(GEOPM_DOMAIN_BOARD)).WillRepeatedly(Return(M_NUM_BOARD));
+    std::vector<double> policy;
+    policy = m_default_policy;
```
Merge these two lines: std::vector policy = m_default_policy

avilcheslopez commented on test/GPUActivityAgentTest.cpp:
```diff
@@ -0,0 +1,434 @@
+        static const int M_NUM_BOARD = 1;
+        static const int M_NUM_GPU = 1;
+        std::unique_ptr<GPUActivityAgent> m_agent;
+        std::vector<double> m_default_policy;
```
Could and should be const

avilcheslopez commented on test/GPUActivityAgentTest.cpp:
```diff
@@ -0,0 +1,434 @@
+
+TEST_F(GPUActivityAgentTest, adjust_platform_high)
+{
+    //TODO: Setup f_max, min etc...via read
```
Is this really necessary?

avilcheslopez commented on test/GPUActivityAgentTest.cpp:
```diff
@@ -0,0 +1,434 @@
+    m_agent->validate_policy(policy);
+
+    //Sample
+    std::vector<double> tmp;
+    double mock_active = 1.0;
+    double mock_util = 1.0;
+    EXPECT_CALL(*m_platform_io, sample(GPU_COMPUTE_ACTIVITY_IDX))
+                .WillRepeatedly(Return(mock_active));
+    EXPECT_CALL(*m_platform_io, sample(GPU_UTILIZATION_IDX))
+                .WillRepeatedly(Return(mock_util));
+    EXPECT_CALL(*m_platform_io, sample(GPU_FREQUENCY_IDX))
+                .WillRepeatedly(Return(m_freq_max - 1)); //Any non-m_freq_max value will work here
+    EXPECT_CALL(*m_platform_io, sample(GPU_ENERGY_IDX))
+                .WillRepeatedly(Return(123456789));
+    m_agent->sample_platform(tmp);
+
+    //Adjust
+    //Check frequency
+    EXPECT_CALL(*m_platform_io, adjust(GPU_FREQUENCY_CONTROL_MIN_IDX, m_freq_max)).Times(M_NUM_GPU);
+    EXPECT_CALL(*m_platform_io, adjust(GPU_FREQUENCY_CONTROL_MAX_IDX, m_freq_max)).Times(M_NUM_GPU);
+    m_agent->adjust_platform(policy);
```
Opportunity to refactor: similar code across all "adjust_platform_*" tests

avilcheslopez commented on test/GPUActivityAgentTest.cpp:
```diff
@@ -0,0 +1,434 @@
+TEST_F(GPUActivityAgentTest, adjust_platform_high)
+{
+    //TODO: Setup f_max, min etc...via read
+    EXPECT_CALL(*m_platform_io, read_signal("GPU_CORE_FREQUENCY_MIN_AVAIL", _, _)).WillRepeatedly(
+                Return(m_freq_min));
+    EXPECT_CALL(*m_platform_io, read_signal("GPU_CORE_FREQUENCY_MAX_AVAIL", _, _)).WillRepeatedly(
+                Return(m_freq_max));
+    EXPECT_CALL(*m_platform_topo, num_domain(GEOPM_DOMAIN_BOARD)).WillRepeatedly(Return(M_NUM_BOARD));
```
This block of code is repeated at the beginning of every test, so it could be moved into the ``SetUp()`` method.

avilcheslopez commented on test/GPUActivityAgentTest.cpp:
```diff
@@ -0,0 +1,434 @@
+using geopm::GPUActivityAgent;
+using geopm::PlatformTopo;
+
+class GPUActivityAgentTest : public :: testing :: Test
```
Additional tests could be added for:
- do_send_policy()
- do_send_sample()
- policy_names()
- report_host()?
- do_write_batch()?

avilcheslopez committed:
Refactor GPUActivityAgent unit tests, other clean-up

- Factor out common code in GPUActivityAgentTest
- Do minor clean-up in documentation
- Remove calls to read signal GPU_CORE_FREQUENCY_STATUS in
  GPUActivityAgent as it's no longer needed
- Do minor code clean-up in GPUActivityAgent

Signed-off-by: Alejandro Vilches <alejandro.vilches@intel.com>

avilcheslopez committed:
Make GPUActivityAgent a beta feature

Signed-off-by: Alejandro Vilches <alejandro.vilches@intel.com>

lhlawson commented on integration/experiment/gpu_frequency_sweep/gen_gpu_activity_policy_recommendation.py:
```diff
@@ -0,0 +1,93 @@
+    # config, ex: GPU_FREQ_MAX: NAN --> HW Maximum Frequency, GPU_PHI --> 0.5
+    # (balanced mode).
+    policy = {"GPU_FREQ_MAX" : float('nan'),
+                "GPU_FREQ_EFFICIENT" : gpu_freq_efficient,
+                "GPU_PHI" : float('nan')}
```
```suggestion
              "GPU_FREQ_EFFICIENT" : gpu_freq_efficient,
              "GPU_PHI" : float('nan')}
```

lhlawson commented on integration/test/test_gpu_activity_agent.py:
```diff
@@ -0,0 +1,138 @@
+
+        max_freq = geopm_test_launcher.geopmread("GPU_CORE_FREQUENCY_MAX_AVAIL board 0")
+
+        # TODO:
+        # Once the GPU Frequency sweep infrastructure is added a full
+        # characterization integration test should be added that does
+        # a GPU frequency sweep of parres dgemm, parses the frequency
+        # sweep using the gen_gpu_activity_policy_recommendation.py
+        # script, and uses the output provided.
+        #
+        # This is a less time consuming version of that approach,
+        # where efficient freq is estimated based on min and max
```
This should be tackled after merging #2024

lhlawson commented on integration/test/test_gpu_activity_agent.py:
```diff
@@ -0,0 +1,138 @@
+@util.skip_unless_config_enable('beta')
+@util.skip_unless_do_launch()
+@util.skip_unless_gpu()
+@util.skip_unless_workload_exists("apps/parres/Kernels/Cxx11/dgemm-mpi-cublas")
+@util.skip_unless_workload_exists("apps/parres/Kernels/Cxx11/nstream-mpi-cuda")
```
If we remove these the test will fail on run instead of skipping.  This might be a better default behavior if we only plan to run this script when the workloads have been built.

avilcheslopez replied:
Note from Brad: leave skips. File issue for feature request: "as a developer of GEOPM I would like to be able to run weekly tests vs. actual workloads in the integration apps directory."

lhlawson committed:
wsp fix integration/experiment/gpu_frequency_sweep/gen_gpu_activity_policy_recommendation.py

wsp fix in recommendation script

lhlawson committed:
Updated to match new parres.py function calls & increase stream iterations

Signed-off-by: lowren.h.lawson@intel.com <lowren.h.lawson@intel.com>

bgeltz commented on Makefile.am:
```diff
@@ -115,6 +115,7 @@ roff_man_inst = man/GEOPM_CXX_MAN_Agent.3 \
                 # end
 
 roff_man_noinst = man/GEOPM_CXX_MAN_Comm.3 \
+                  man/GEOPM_CXX_MAN_GPUActivityAgent.3 \
```
Any man pages we need to add should also be behind the beta flag.  This should removed, and the chapter 7 included in the ```roff_man_beta``` section.

bgeltz commented on service/docs/source/geopm_agent_gpu_activity.7.rst:
```diff
@@ -0,0 +1,162 @@
+
+    {"GPU_FREQ_MAX": NaN,
+     "GPU_FREQ_EFFICIENT": 982000000.0,
+      "GPU_PHI": NaN}
```
```suggestion
     "GPU_PHI": NaN}
```
Off by 1

bgeltz commented on integration/experiment/energy_efficiency/run_gpu_activity_parres_dgemm.py:
```diff
@@ -0,0 +1,24 @@
+    mach = machine.init_output_dir(args.output_dir)
+    app_conf = parres.create_dgemm_appconf(mach, args)
+    gpu_activity.launch(app_conf=app_conf, args=args,
+                   experiment_cli_args=extra_args)
```
Indentation issue

bgeltz commented on integration/experiment/energy_efficiency/run_gpu_activity_parres_nstream.py:
```diff
@@ -0,0 +1,24 @@
+    mach = machine.init_output_dir(args.output_dir)
+    app_conf = parres.create_nstream_appconf(mach, args)
+    gpu_activity.launch(app_conf=app_conf, args=args,
+                   experiment_cli_args=extra_args)
```
Indentation issue

bgeltz commented on src/GPUActivityAgent.hpp:
```diff
@@ -0,0 +1,110 @@
+            bool m_do_write_batch;
+            bool m_do_send_policy;
+
+            struct signal
```
```suggestion
            struct m_signal
```

bgeltz commented on src/GPUActivityAgent.hpp:
```diff
@@ -0,0 +1,110 @@
+                double value;
+            };
+
+            struct control
```
```suggestion
            struct m_control
```
We should've done this for the CPU-CA too.

bgeltz commented on test/GPUActivityAgentTest.cpp:
```diff
@@ -0,0 +1,319 @@
+#include "config.h"
+
+#include <stdlib.h>
+#include <iostream>
+#include <fstream>
```
```suggestion
```
Can these all be removed?  If not, remove as much as possible.

bgeltz commented on src/GPUActivityAgent.cpp:
```diff
@@ -0,0 +1,434 @@
+    }
+
+    void GPUActivityAgent::aggregate_sample(const std::vector<std::vector<double> > &in_sample,
+                                        std::vector<double>& out_sample)
```
Indentation

lhlawson committed:
Addressed review feedback: indentation, struct name update

Signed-off-by: lowren.h.lawson@intel.com <lowren.h.lawson@intel.com>

lhlawson committed:
Updated GPU-CA unit tests to remove constexpr

Signed-off-by: lowren.h.lawson@intel.com <lowren.h.lawson@intel.com>

dannosliwcd commented on integration/experiment/gpu_frequency_sweep/gen_gpu_activity_policy_recommendation.py:
```diff
@@ -0,0 +1,93 @@
+        sys.exit(1)
+
+    output = main(df)
+    sys.stdout.write("AGENT POLICY:\n")
```
```suggestion
```
If the tool only prints the generated policy to stdout, then users can `./gen_gpu_activity_policy_recommendation.py > /path/to/policy/file`

dannosliwcd replied:
Could alternatively add a `--output=/path/to/output/file` option. But for a tool that generates the contents of only  one file, `generate > file` feels more typical.

avilcheslopez committed:
Fix GPUActivityAgentTest issue with clang

Signed-off-by: Alejandro Vilches <alejandro.vilches@intel.com>

lhlawson commented on integration/experiment/energy_efficiency/run_gpu_activity_parres_dgemm.py:
```diff
@@ -0,0 +1,24 @@
+    parres.setup_run_args(parser)
+    args, extra_args = parser.parse_known_args()
+    mach = machine.init_output_dir(args.output_dir)
+    app_conf = parres.create_dgemm_appconf(mach, args)
```
```suggestion
    app_conf = parres.create_dgemm_cuda_appconf(mach, args)
```

lhlawson commented on integration/experiment/energy_efficiency/run_gpu_activity_parres_dgemm.py:
```diff
@@ -0,0 +1,24 @@
+from experiment.energy_efficiency import gpu_activity
+from apps.parres import parres
+
+
```
rename file to run_gpu_activity_parres_dgemm_cuda.py

lhlawson commented on integration/experiment/energy_efficiency/run_gpu_activity_parres_nstream.py:
```diff
@@ -0,0 +1,24 @@
+from experiment.energy_efficiency import gpu_activity
+from apps.parres import parres
+
+
```
rename file to run_gpu_activity_parres_nstream_cuda.py

lhlawson commented on integration/experiment/energy_efficiency/run_gpu_activity_parres_nstream.py:
```diff
@@ -0,0 +1,24 @@
+    parres.setup_run_args(parser)
+    args, extra_args = parser.parse_known_args()
+    mach = machine.init_output_dir(args.output_dir)
+    app_conf = parres.create_nstream_appconf(mach, args)
```
```suggestion
    app_conf = parres.create_nstream_cuda_appconf(mach, args)
```

lhlawson committed:
updated GPU-CA experiment scripts to new naming scheme

Signed-off-by: lowren.h.lawson@intel.com <lowren.h.lawson@intel.com>

bgeltz commented on integration/experiment/energy_efficiency/Makefile.mk:
```diff
@@ -17,6 +18,8 @@ EXTRA_DIST += integration/experiment/energy_efficiency/barrier_frequency_sweep.p
               integration/experiment/energy_efficiency/run_cpu_activity_nasft.py \
               integration/experiment/energy_efficiency/run_cpu_activity_nekbone.py \
               integration/experiment/energy_efficiency/run_cpu_activity_pennant.py \
+              integration/experiment/energy_efficiency/run_gpu_activity_parres_dgemm_cuda.py \
+              integration/experiment/energy_efficiency/run_gpu_activity_parres_nstream_cuda.py \
```
```suggestion

              integration/experiment/energy_efficiency/run_gpu_activity_parres_dgemm_cuda.py \
              integration/experiment/energy_efficiency/run_gpu_activity_parres_nstream_cuda.py \
                            integration/experiment/energy_efficiency/run_cpu_activity_pennant.py \
```

bgeltz commented on integration/experiment/Makefile.mk:
```diff
@@ -21,4 +21,5 @@ include integration/experiment/frequency_sweep/Makefile.mk
 include integration/experiment/monitor/Makefile.mk
 include integration/experiment/power_sweep/Makefile.mk
 include integration/experiment/trace_analysis/Makefile.mk
+include integration/experiment/gpu_frequency_sweep/Makefile.mk
```
```suggestion
include integration/experiment/monitor/Makefile.mk
include integration/experiment/gpu_frequency_sweep/Makefile.mk
include integration/experiment/power_sweep/Makefile.mk
include integration/experiment/trace_analysis/Makefile.mk
```

bgeltz commented on copying_headers/MANIFEST.EXEMPT:
```diff
@@ -62,6 +62,7 @@ integration/experiment/power_sweep/README.md
 integration/experiment/README.md
 integration/experiment/trace_analysis/README.md
 integration/experiment/uncore_frequency_sweep/README.md
+integration/experiment/gpu_frequency_sweep/README.md
```
Move this up under the monitor.  Please

bgeltz commented on src/GPUActivityAgent.cpp:
```diff
@@ -0,0 +1,434 @@
+                //
+                // If f_efficient does not meet these criteria this behavior can still be
+                // achieved through tracking the GPU Utilization signal and setting frequency
+                // to a separate idle value (f_idle) during regions where GPU Utilizaiton is
```
```suggestion
                // to a separate idle value (f_idle) during regions where GPU Utilization is
```

lhlawson committed:
Addressed review feedback on man pages and re-ordering includes

Signed-off-by: lowren.h.lawson@intel.com <lowren.h.lawson@intel.com>

lhlawson committed:
Add activity agents documentation to Makefile

Signed-off-by: lowren.h.lawson@intel.com <lowren.h.lawson@intel.com>

bgeltz committed:
Fix the build

Signed-off-by: Brad Geltz <brad.geltz@intel.com>
Change-Id: I18bb46c4dbded9cb9dc140e0534526b24bec63ba
